### ISSUE TITLE
usnic: Support of synchronous fi_cq_sread?

### ISSUE THREAD

**Issue by wavesoft** [ofiwg/libfabric#1629]
I recently stumbled into a problem, using the `libfabric-1.2.0` release and the `usnic` provider.

So far, I am polling, waiting for the Rx/Tx event to appear in the CQ [as seen here](https://github.com/wavesoft/nanomsg-transport-ofi/blob/master/src/transports/ofi/hlapi.c#L202), but this introduces some inconsistent latency measurements on my benchmarks.
I did some tests with blocking (synchronous) calls, to the `fi_cq_sread` function and on the `sockets` provider and it seems to solve my problem. However, as [i confirm from the sources](https://github.com/ofiwg/libfabric/blob/master/prov/usnic/src/usdf_cq.c#L1105), `usnic` provider does not support (yet?) wait objects on CQs.

So, I was wondering:
 1. Is this going to be implemented soon? Is there any planned milestone/date?
 2. In the mean time, is there any other recommended way to get a faster and more reliable notification?

**Comment by jsquyres** [comment]
@wavesoft Sorry for the delay in replying.  Yes, we're literally in the middle of implementing fi_waitset support for usNIC right now (we need reliable / blocking notifications ourselves for other usNIC-based projects).

We hope to include fi_waitset support in libfabric v1.3.0.  However, fi_waitset support will *also* require a VIC firmware update to handle the interrupt generation support.  The firmware will likely take a while to become available (we can push libfabric code up here to Github to make it available to the community, but firmware releases go through the normal develop/QA/release trains for all cisco.com software releases).

**Comment by wavesoft** [comment]
@jsquyres Thanks for the update! That's very useful information. I will wait for the 1.3.0 release then. I assume it will come on best-effort? Or is there any scheduled timeline?

In the meantime I will check if here at CERN we can get an earlier access to CISCO firmware, through the OpenLab collaboration.

**Comment by jsquyres** [comment]
@wavesoft We definitely need fi_waitset support for other stuff, so I would *hope* that we can get it all done / tested / included in libfabric v1.3.0.  I cannot definitely promise that, however (assuming that the v1.3.0 deadline is March).

**Comment by goodell** [comment]
@wavesoft There are two issues here that we should address separately:

1. waitset support in the usnic provider
2. `fi_cq_sread` support in the usnic provider

The first isn't yet present, but Jeff has discussed the issue above.

The second is and has been well supported since the v1.2.0 release (there was a performance issue prior to that release).  When I look at [your linked code](https://github.com/wavesoft/nanomsg-transport-ofi/blob/master/src/transports/ofi/hlapi.c#L205-L207), I don't see what problem you would have from uncommenting the `fi_cq_sread` line.  Can you elaborate what sort of issue you're having with `fi_cq_sread` and the usnic provider?

**Comment by wavesoft** [comment]
@goodell Sure, the problem is not on the `fi_cq_sread` function, rather on the fact that the `fi_cq_open` function ([located here in my code](https://github.com/wavesoft/nanomsg-transport-ofi/blob/master/src/transports/ofi/hlapi.c#L556)) does [not support a wait object](https://github.com/ofiwg/libfabric/blob/master/prov/usnic/src/usdf_cq.c#L1105).

I don't think it makes sense to call the `fi_cq_sread` function without defining a wait object, right?

**Comment by wavesoft** [comment]
Ah I see your point, the title might be a bit misleading. I was referring to your first point. The waitset support on `usnic` provider.

**Comment by goodell** [comment]
> I don't think it makes sense to call the fi_cq_sread function without defining a wait object, right?

It makes plenty of sense, depending on your application usage.  A wait object can be thought of as a generalization of a file descriptor, like you would use to `poll` or `select` when writing sockets code.  `fi_cq_sread` is just supposed to be equivalent to an optimized version of:

```c
do {
    ret = fi_cq_read(...);
    if (ret == -FI_EAGAIN)
        do_sleep();
} while (ret == -FI_EAGAIN && timeout_ms > 0);
```

Where "optimized" means uses less CPU or other hardware resources, or possibly gives you a better response time compared to a naive scheme.

The point where you would really need waitset support is if you have multiple CQs that you want to block on.  You can share a single waitset among multiple CQs, whereas I don't think there's a way to `fi_cq_sread` multiple CQs at the same time without using your own thread for each CQ.

**Comment by wavesoft** [comment]
Perhaps I was not very clear... I mean that if I specify `.wait_obj = FI_WAIT_NONE` during the creation of the CQ, it makes no sense to call the `fi_cq_sread` function. I might be wrong, but I am quoting this form the `fi_cq` manpage:

> FI_WAIT_NONE
>    Used to indicate that the user will not block (wait) for completions on the CQ. When FI_WAIT_NONE  is specified, the application may not call fi_cq_sread or fi_cq_sreadfrom.

**Comment by goodell** [comment]
Hmm... I never noticed that in the man page.  I don't agree with the statement

> if I specify .wait_obj = FI_WAIT_NONE during the creation of the CQ, it makes no sense to call the fi_cq_sread function

in general.  I don't feel like one should be needed for a synchronous wait, since a provider can always provide a loop-based fallback implementation for `fi_cq_sread` if the user doesn't provide a wait object at CQ creation time.

@shefty I have a strong suspicion that you wrote this portion of the man page.  Can you shed some light on the reasoning here?  I thought wait objects were mainly intended for multiplexing multiple event sources into a single sleepable object or for interoperating cleanly with OS resources (e.g., other FDs from regular sockets/files).

**Comment by shefty** [comment]
> @shefty <https://github.com/shefty>  I have a strong suspicion that you
> wrote this portion of the man page. Can you shed some light on the
> reasoning here? I thought wait objects were mainly intended for
> multiplexing multiple event sources into a single sleepable object or for
> interoperating cleanly with OS resources (e.g., other FDs from regular
> sockets/files).

FI_WAIT_NONE is a way for the app to indicate that it will never make a blocking call or try to wait for a completion.  This is different than FI_WAIT_UNSPEC, which says that it will, but doesn't care how the wait is implemented.  Wait *objects* are intended for blocking until some event occurs.  Wait *sets* are intended for multiplexing multiple event sources.  If the app cares about interoperating with OS resources, they need to specify a specific type of wait object.

For software based solutions, wait objects and wait sets may essentially be the same.  For example, the sockets and PSM (and soon to be UDP) providers implement waiting by allocating wait sets internally.  Hardware based solutions (such as verbs) may not have this flexibility.

**Comment by goodell** [comment]
@shefty fair enough, I'll tweak the usnic provider to make this NONE/UNSPEC distinction.

**Comment by jsquyres** [comment]
This issue was addressed in PR #1639.
